eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs

```diff
@@ -156,24 +175,22 @@ collectTwoPhaseScriptInputs ::
   UTxO era ->
   Either [CollectError (Crypto era)] [(ShortByteString, Language, [Data era], ExUnits, CostModel)]
 collectTwoPhaseScriptInputs ei sysS pp tx utxo =
-  let usedLanguages = [lang | (AlonzoScript.PlutusScript lang _) <- Map.elems scriptsUsed]
+  let usedLanguages = [lang | (_, lang, _) <- neededAndConfirmedToBePlutus]
       costModels = unCostModels $ getField @"_costmdls" pp
       missingCMs = [lang | lang <- usedLanguages, lang `Map.notMember` costModels]
    in case missingCMs of
         l : _ -> Left [NoCostModel l]
-        _ -> merge (apply costModels) (map redeemer needed) (map getscript needed) (Right [])
+        _ ->
```

Comment:
I just realized the whole thing should be made more efficient:
```haskell
  let usedLanguages = Set.fromList [lang | (_, lang, _) <- neededAndConfirmedToBePlutus]
      costModels = unCostModels $ getField @"_costmdls" pp
      missingCMs = Set.filter (`Map.notMember` costModels) usedLanguages
   in case Set.lookupMin missingCMs of
        Just l -> Left [NoCostModel l]
        Nothing -> ...
```

Reply:
thanks!

Reply:
done